Make `as_nonzero_unchecked()` a const fn?

the `new_unchecked()` fn on the various NonZero types is a const fn, so I think if `as_nonzero_unchecked()` was as well, it would be possible to write things like:

```rust
const MY_NON_ZERO_U32: NonZeroU32 = nonzero!(42u32);
```

currently it gives the following compile error:

```
error[E0015]: calls in constants are limited to constant functions, tuple structs and tuple variants
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "nonzero_ext"
-version = "0.1.6-dev"
+version = "0.2.0-dev"
 authors = ["Andreas Fuchs <asf@boinkor.net>"]
 license = "Apache-2.0"
 repository = "https://github.com/antifuchs/nonzero_ext"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -95,6 +95,8 @@ mod lib {
 
 use self::lib::*;
 
+pub mod literals;
+
 macro_rules! impl_nonzeroness {
     ($trait_name:ident, $nonzero_type:ty, $wrapped:ty) => {
         impl $trait_name for $nonzero_type {
@@ -198,6 +200,11 @@ macro_rules! impl_nonzeroable {
                 Self::NonZero::new_unchecked(self)
             }
         }
+        impl literals::NonZeroLiteral<$nonzeroable_type> {
+            pub const unsafe fn into_nonzero(self) -> $nonzero_type {
+                <$nonzero_type>::new_unchecked(self.0)
+            }
+        }
     };
 }
 
@@ -223,6 +230,11 @@ impl_nonzeroable!(NonZeroAble, NonZeroIsize, isize);
 /// known, `nonzero!` requires that you annotate the constant with the
 /// type, so instead of `nonzero!(20)` you must write `nonzero!(20 as
 /// u16)`.
+///
+/// # Const expressions
+///
+/// This macro can be used in const expressions.
+///
 /// # Examples
 /// ```
 /// # #[macro_use]
@@ -247,12 +259,9 @@ impl_nonzeroable!(NonZeroAble, NonZeroIsize, isize);
 #[macro_export]
 macro_rules! nonzero {
     ($n:expr) => {{
-        let helper = || {
-            #[allow(unknown_lints, eq_op)]
-            let _ = [(); ($n.count_ones() as usize) - 1];
-            use $crate::NonZeroAble;
-            unsafe { $n.as_nonzero_unchecked() }
-        };
-        helper()
+        #[allow(unknown_lints, eq_op)]
+        let _ = [(); ($n.count_ones() as usize) - 1];
+        let lit = $crate::literals::NonZeroLiteral($n);
+        unsafe { lit.into_nonzero() }
     }};
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -201,6 +201,9 @@ macro_rules! impl_nonzeroable {
             }
         }
         impl literals::NonZeroLiteral<$nonzeroable_type> {
+            /// Converts the wrapped value to its non-zero equivalent.
+            /// # Safety
+            /// The wrapped value must be non-zero.
             pub const unsafe fn into_nonzero(self) -> $nonzero_type {
                 <$nonzero_type>::new_unchecked(self.0)
             }

--- a/src/literals.rs
+++ b/src/literals.rs
@@ -1,0 +1,9 @@
+//! Handling non-zero literal values.
+
+use super::NonZeroAble;
+
+/// A representation of a non-zero literal. Used by the [`nonzero!`] macro.
+///
+/// This struct has no use outside of this macro (even though it can be constructed by anyone).
+/// It needs to exist to support the use of the [`nonzero!`] macro in const expressions.
+pub struct NonZeroLiteral<T: NonZeroAble>(pub T);

--- a/tests/compile-fail/immoral_math.rs
+++ b/tests/compile-fail/immoral_math.rs
@@ -1,0 +1,8 @@
+#[macro_use]
+extern crate nonzero_ext;
+
+use std::num::NonZeroU32;
+
+fn main() {
+    const _MY_NON_ZERO_U32: NonZeroU32 = nonzero!((42 / 2 - 21) as u32);
+}

--- a/tests/compile-fail/immoral_math.stderr
+++ b/tests/compile-fail/immoral_math.stderr
@@ -1,0 +1,7 @@
+error[E0080]: evaluation of constant value failed
+ --> $DIR/immoral_math.rs:7:42
+  |
+7 |     const _MY_NON_ZERO_U32: NonZeroU32 = nonzero!((42 / 2 - 21) as u32);
+  |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to subtract with overflow
+  |
+  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

--- a/tests/run-pass/const_expression.rs
+++ b/tests/run-pass/const_expression.rs
@@ -1,0 +1,8 @@
+#[macro_use]
+extern crate nonzero_ext;
+
+use std::num::NonZeroU32;
+
+fn main() {
+    const _MY_NON_ZERO_U32: NonZeroU32 = nonzero!(42u32);
+}

--- a/tests/run-pass/virtuous_math.rs
+++ b/tests/run-pass/virtuous_math.rs
@@ -1,0 +1,8 @@
+#[macro_use]
+extern crate nonzero_ext;
+
+use std::num::NonZeroU32;
+
+fn main() {
+    const _MY_NON_ZERO_U32: NonZeroU32 = nonzero!((42 * 2 - 21) as u32);
+}

